mungegithub/mungers/submit-queue.go

```diff
@@ -927,110 +933,138 @@ func (sq *SubmitQueue) handleGithubE2EAndMerge() {
 			continue
 		}
 
-		sq.Lock()
-		if len(sq.githubE2EQueue) == 0 {
-			sq.Unlock()
+		obj := sq.selectPullRequest()
+		if obj == nil {
 			continue
 		}
-		keys := sq.orderedE2EQueue()
-		obj := sq.githubE2EQueue[keys[0]]
-		sq.githubE2ERunning = obj
-		sq.Unlock()
 
 		// re-test and maybe merge
-		sq.doGithubE2EAndMerge(obj)
+		if sq.doGithubE2EAndMerge(obj) {
+			// remove it from the map after we finish testing
+			sq.Lock()
+			sq.githubE2ERunning = nil
+			delete(sq.githubE2EQueue, *obj.Issue.Number)
+			sq.Unlock()
+		}
+	}
+}
 
-		// remove it from the map after we finish testing
-		sq.Lock()
-		sq.githubE2ERunning = nil
-		delete(sq.githubE2EQueue, keys[0])
-		sq.Unlock()
+func (sq *SubmitQueue) mergePullRequest(obj *github.MungeObject) {
+	obj.MergePR("submit-queue")
+	sq.updateMergeRate()
+	sq.SetMergeStatus(obj, merged)
+}
+
+func (sq *SubmitQueue) reportFailedPullRequest(obj *github.MungeObject, err string) {
```

Comment:
No it's useless. I was focused on success/failure symmetry. I didn't realize this was useless ... Will remove!